File: solvers/ac3/heuristics.py
"""Variable / value ordering for Algo3 - unchanged from Algo3/heuristics.py."""
from .utils import number_of_conflicts


def select_unassigned_variable(assignment, sudoku):
    """MRV: the unassigned cell with the fewest remaining possibilities."""
    unassigned = []
    for cell in sudoku.cells:
        if cell not in assignment:
            unassigned.append(cell)
    criterion = lambda cell: len(sudoku.possibilities[cell])
    return min(unassigned, key=criterion)


def order_domain_values(sudoku, cell):
    """LCV: values that rule out the fewest neighbour choices, first."""
    if len(sudoku.possibilities[cell]) == 1:
        return sudoku.possibilities[cell]
    criterion = lambda value: number_of_conflicts(sudoku, cell, value)
    return sorted(sudoku.possibilities[cell], key=criterion)

File: web/packages/solver-core/src/solvers/ac3/heuristics.ts
/** Port of solvers/ac3/heuristics.py (MRV variable order, LCV value order). */
import type { Sudoku } from "./sudoku.js";
import { numberOfConflicts } from "./utils.js";

export function selectUnassignedVariable(
  assignment: Map<string, number>,
  sudoku: Sudoku,
): string {
  const unassigned: string[] = [];
  for (const cell of sudoku.cells) {
    if (!assignment.has(cell)) unassigned.push(cell);
  }
  // Python min(unassigned, key=len): first cell achieving the minimum.
  let best = unassigned[0];
  let bestLen = sudoku.possibilities.get(best)!.length;
  for (let i = 1; i < unassigned.length; i++) {
    const len = sudoku.possibilities.get(unassigned[i])!.length;
    if (len < bestLen) {
      best = unassigned[i];
      bestLen = len;
    }
  }
  return best;
}

export function orderDomainValues(sudoku: Sudoku, cell: string): number[] {
  const poss = sudoku.possibilities.get(cell)!;
  if (poss.length === 1) return poss;
  // Python sorted() is stable; so is V8's Array.prototype.sort.
  return [...poss].sort(
    (a, b) => numberOfConflicts(sudoku, cell, a) - numberOfConflicts(sudoku, cell, b),
  );
}
